Remove VideoToolbox VP8 and VP9 plugins

jyavenard opened the pull request:
#### c077ba3cb140797d981cd48e3c37f4e55e9efcaf
<pre>
Remove VideoToolbox VP8 and VP9 plugins
<a href="https://bugs.webkit.org/show_bug.cgi?id=282163">https://bugs.webkit.org/show_bug.cgi?id=282163</a>
<a href="https://rdar.apple.com/138751645">rdar://138751645</a>

Reviewed by Youenn Fablet.

The VideoToolbox VP8 and VP plugins were only usable on Macs and were non-functional on other platforms which
use mediaserverd.
We can streamline the code path and use the WebCodecs&apos; VideoDecoder code path on all platforms; doing so make
the VideoToolbox plugins dead-code.
While we had a preference to disable the VP8 settings, it was in effect non-functional on WK2 as we would always default
to the VideoDecoder if the VideoToolbox was disabled. So we can remove this setting altogether considering the time
we have shipped with an always enabled VP8 decoder.
We remove the settings. A side effect is that it will now enable VP8 on WK1 however considering:
- We reduce the number of diverging code path.
- WK1 had support for WebM with Vorbis and Opus, just not VP8.
  This made WebM in a semi-broken state as VP8 is the most common codec found with this container
it&apos;s an acceptable compromise.

Covered by existing tests.

* LayoutTests/media/media-sources-selection.html:
* LayoutTests/platform/mac-wk1/TestExpectations: Re-enable MSE tests with webm using vp8 videos.
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/media-source/mediasource-addsourcebuffer-expected.txt: vp8 is now supported, update expectation.
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/media-source/SourceBuffer-abort-updating-expected.txt: Removed.
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/media-source/mediasource-invalid-codec-expected.txt: Removed.
* PerformanceTests/DecoderTest/DecoderTest/main.mm:
(main):
* Source/ThirdParty/libwebrtc/CMakeLists.txt:
* Source/ThirdParty/libwebrtc/Configurations/libwebrtc.exp:
* Source/ThirdParty/libwebrtc/Source/webrtc/webkit_sdk/WebKit/WebKitDecoderReceiver.cpp: Removed.
* Source/ThirdParty/libwebrtc/Source/webrtc/webkit_sdk/WebKit/WebKitDecoderReceiver.h: Removed.
* Source/ThirdParty/libwebrtc/Source/webrtc/webkit_sdk/WebKit/WebKitVP8Decoder.cpp: Removed.
* Source/ThirdParty/libwebrtc/Source/webrtc/webkit_sdk/WebKit/WebKitVP8Decoder.h: Removed.
* Source/ThirdParty/libwebrtc/Source/webrtc/webkit_sdk/WebKit/WebKitVP9Decoder.cpp: Removed.
* Source/ThirdParty/libwebrtc/Source/webrtc/webkit_sdk/WebKit/WebKitVP9Decoder.h: Removed.
* Source/ThirdParty/libwebrtc/libwebrtc.xcodeproj/project.pbxproj:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/platform/VideoDecoder.cpp:
(WebCore::VideoDecoder::isVPXSupported):
* Source/WebCore/platform/VideoDecoder.h:
* Source/WebCore/platform/audio/PlatformMediaSessionManager.cpp:
(WebCore::PlatformMediaSessionManager::setShouldEnableVP8Decoder): Deleted.
(WebCore::PlatformMediaSessionManager::shouldEnableVP8Decoder): Deleted.
* Source/WebCore/platform/audio/PlatformMediaSessionManager.h:
* Source/WebCore/platform/audio/cocoa/MediaSessionManagerCocoa.mm:
(WebCore::MediaSessionManagerCocoa::ensureCodecsRegistered):
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
(WebCore::MediaPlayerPrivateWebM::seekInternal): Fly-By: no need to flush the tracks again, as we just flushed everything earlier.
* Source/WebCore/platform/graphics/cocoa/VP9UtilitiesCocoa.h:
* Source/WebCore/platform/graphics/cocoa/VP9UtilitiesCocoa.mm:
(WebCore::isVP9DecoderAvailable):
(WebCore::isVP8DecoderAvailable):
(WebCore::registerWebKitVP9Decoder): Deleted.
(WebCore::registerWebKitVP8Decoder): Deleted.
* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
(WebCore::VideoMediaSampleRenderer::enqueueSample):
* Source/WebCore/platform/mediastream/libwebrtc/LibWebRTCProvider.cpp:
(WebCore::LibWebRTCProvider::registerWebKitVP9Decoder): Deleted.
(WebCore::LibWebRTCProvider::registerWebKitVP8Decoder): Deleted.
* Source/WebCore/platform/mediastream/libwebrtc/LibWebRTCProvider.h:
* Source/WebCore/platform/mediastream/libwebrtc/LibWebRTCProviderCocoa.cpp:
(WebCore::LibWebRTCProvider::registerWebKitVP9Decoder): Deleted.
(WebCore::LibWebRTCProvider::registerWebKitVP8Decoder): Deleted.
* Source/WebKit/GPUProcess/GPUProcess.cpp:
(WebKit::GPUProcess::updateGPUProcessPreferences):
* Source/WebKit/GPUProcess/GPUProcess.h:
* Source/WebKit/GPUProcess/GPUProcessPreferences.cpp:
(WebKit::GPUProcessPreferences::copyEnabledWebPreferences):
* Source/WebKit/GPUProcess/GPUProcessPreferences.h:
* Source/WebKit/GPUProcess/GPUProcessPreferences.serialization.in:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::creationParameters):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_textAnimationController):

Canonical link: <a href="https://commits.webkit.org/287137@main">https://commits.webkit.org/287137@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b15cabcf8d6a709c0c33d54501108b1b2aba1b8f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/78481 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/57526 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/31863 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/83142 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/29746 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/80614 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/66677 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/5807 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/61466 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/19385 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/81548 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/51481 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/69406 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/41778 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/48827 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/25251 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/28083 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/71627 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/69925 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/25623 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/84508 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/77718 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/5846 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/3989 "Found 1 new test failure: imported/w3c/web-platform-tests/html/browsers/the-window-object/open-close/open-features-tokenization-noreferrer.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/69692 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/6007 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/67467 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/68947 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17182 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/12969 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/11364 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/100027 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/5793 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/9195 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/21844 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/5781 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/9215 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/7568 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/platform/graphics/cocoa/VP9UtilitiesCocoa.mm:
```diff
@@ -155,23 +145,12 @@ bool isVP9DecoderAvailable()
-    if (!VideoToolboxLibrary(true))
-        return false;
-    return noErr == VTSelectAndCreateVideoDecoderInstance(kCMVideoCodecType_VP9, kCFAllocatorDefault, nullptr, nullptr);
+    return true;
```
When libwebrtc is not available, we do not have access to libvpx.
We should keep those checks and if not successful, check for libwebrtc being available.

jyavenard replied:
then decoding will error.

jyavenard replied:
I've amended the test, now checking if the VideoDecoder supports vp8 or vp9

youennf commented on Source/WebCore/platform/graphics/cocoa/VP9UtilitiesCocoa.mm:
```diff
@@ -155,23 +145,12 @@ bool isVP9DecoderAvailable()
-        return true;
-    if (!VideoToolboxLibrary(true))
-        return false;
-    return noErr == VTSelectAndCreateVideoDecoderInstance('vp08', kCFAllocatorDefault, nullptr, nullptr);
```
Ditto, we need to check that libvpx/libwebrtc is available, WebRTCProvider::webRTCAvailable.

jyavenard replied:
fine

youennf commented on Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
```diff
@@ -71,12 +72,12 @@ - (void)resetUpcomingSampleBufferPresentationTimeExpectations;
         CMVideoFormatDescriptionRef videoFormatDescription = PAL::CMSampleBufferGetFormatDescription(sample);
         auto fourCC = PAL::CMFormatDescriptionGetMediaSubType(videoFormatDescription);
-        if (fourCC == 'vp08' || (fourCC == 'vp09' && !(canLoad_VideoToolbox_VTIsHardwareDecodeSupported() && VTIsHardwareDecodeSupported(kCMVideoCodecType_VP9))))
+        if (fourCC == 'vp08' || (fourCC == 'vp09' && !vp9HardwareDecoderAvailable()))
```
Looks ok to me, but please check that it is ok to check the testing override here.

jyavenard replied:
we should be consistent across all tests if a decoder is enabled or not.

youennf reviewed (approved):
Should we also remove the VP9DecoderEnabled flag?

jernoble commented on Source/WebCore/platform/libwebrtc/LibWebRTCVPXVideoDecoder.cpp:
```diff
@@ -169,7 +171,9 @@ LibWebRTCVPXInternalVideoDecoder::LibWebRTCVPXInternalVideoDecoder(LibWebRTCVPXV
     : m_outputCallback(WTFMove(outputCallback))
     , m_internalDecoder(createInternalDecoder(type))
     , m_useIOSurface(config.pixelBuffer == VideoDecoder::HardwareBuffer::Yes)
+    , m_isFullRange((type == LibWebRTCVPXVideoDecoder::Type::VP8 || type == LibWebRTCVPXVideoDecoder::Type::VP9 || type == LibWebRTCVPXVideoDecoder::Type::VP9_P2) && config.description.size() >= 12 && (*(config.description.data() + 6) & 0x1))
```
It's deeply weird to add parsing of the VPx header here. What if you added the following to `VideoDecoder::Config`:

```c++
std::optional<bool> isFullRange { };
```

And passed in that information as part of the `VideoDecoder::create()` method?

jyavenard replied:
This is not part of this commit, but a stack of a now obsolete one which you commented earlier on and got addressed.

jyavenard committed:
Remove VideoToolbox VP8 and VP9 plugins
https://bugs.webkit.org/show_bug.cgi?id=282163
rdar://138751645

Reviewed by Youenn Fablet.

The VideoToolbox VP8 and VP plugins were only usable on Macs and were non-functional on other platforms which
use mediaserverd.
We can streamline the code path and use the WebCodecs' VideoDecoder code path on all platforms; doing so make
the VideoToolbox plugins dead-code.
While we had a preference to disable the VP8 settings, it was in effect non-functional on WK2 as we would always default
to the VideoDecoder if the VideoToolbox was disabled. So we can remove this setting altogether considering the time
we have shipped with an always enabled VP8 decoder.
We remove the settings. A side effect is that it will now enable VP8 on WK1 however considering:
- We reduce the number of diverging code path.
- WK1 had support for WebM with Vorbis and Opus, just not VP8.
  This made WebM in a semi-broken state as VP8 is the most common codec found with this container
it's an acceptable compromise.

Covered by existing tests.

* LayoutTests/media/media-sources-selection.html:
* LayoutTests/platform/mac-wk1/TestExpectations: Re-enable MSE tests with webm using vp8 videos.
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/media-source/mediasource-addsourcebuffer-expected.txt: vp8 is now supported, update expectation.
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/media-source/SourceBuffer-abort-updating-expected.txt: Removed.
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/media-source/mediasource-invalid-codec-expected.txt: Removed.
* PerformanceTests/DecoderTest/DecoderTest/main.mm:
(main):
* Source/ThirdParty/libwebrtc/CMakeLists.txt:
* Source/ThirdParty/libwebrtc/Configurations/libwebrtc.exp:
* Source/ThirdParty/libwebrtc/Source/webrtc/webkit_sdk/WebKit/WebKitDecoderReceiver.cpp: Removed.
* Source/ThirdParty/libwebrtc/Source/webrtc/webkit_sdk/WebKit/WebKitDecoderReceiver.h: Removed.
* Source/ThirdParty/libwebrtc/Source/webrtc/webkit_sdk/WebKit/WebKitVP8Decoder.cpp: Removed.
* Source/ThirdParty/libwebrtc/Source/webrtc/webkit_sdk/WebKit/WebKitVP8Decoder.h: Removed.
* Source/ThirdParty/libwebrtc/Source/webrtc/webkit_sdk/WebKit/WebKitVP9Decoder.cpp: Removed.
* Source/ThirdParty/libwebrtc/Source/webrtc/webkit_sdk/WebKit/WebKitVP9Decoder.h: Removed.
* Source/ThirdParty/libwebrtc/libwebrtc.xcodeproj/project.pbxproj:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/platform/VideoDecoder.cpp:
(WebCore::VideoDecoder::isVPXSupported):
* Source/WebCore/platform/VideoDecoder.h:
* Source/WebCore/platform/audio/PlatformMediaSessionManager.cpp:
(WebCore::PlatformMediaSessionManager::setShouldEnableVP8Decoder): Deleted.
(WebCore::PlatformMediaSessionManager::shouldEnableVP8Decoder): Deleted.
* Source/WebCore/platform/audio/PlatformMediaSessionManager.h:
* Source/WebCore/platform/audio/cocoa/MediaSessionManagerCocoa.mm:
(WebCore::MediaSessionManagerCocoa::ensureCodecsRegistered):
* Source/WebCore/platform/graphics/cocoa/MediaPlayerPrivateWebM.mm:
(WebCore::MediaPlayerPrivateWebM::seekInternal): Fly-By: no need to flush the tracks again, as we just flushed everything earlier.
* Source/WebCore/platform/graphics/cocoa/VP9UtilitiesCocoa.h:
* Source/WebCore/platform/graphics/cocoa/VP9UtilitiesCocoa.mm:
(WebCore::isVP9DecoderAvailable):
(WebCore::isVP8DecoderAvailable):
(WebCore::registerWebKitVP9Decoder): Deleted.
(WebCore::registerWebKitVP8Decoder): Deleted.
* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
(WebCore::VideoMediaSampleRenderer::enqueueSample):
* Source/WebCore/platform/mediastream/libwebrtc/LibWebRTCProvider.cpp:
(WebCore::LibWebRTCProvider::registerWebKitVP9Decoder): Deleted.
(WebCore::LibWebRTCProvider::registerWebKitVP8Decoder): Deleted.
* Source/WebCore/platform/mediastream/libwebrtc/LibWebRTCProvider.h:
* Source/WebCore/platform/mediastream/libwebrtc/LibWebRTCProviderCocoa.cpp:
(WebCore::LibWebRTCProvider::registerWebKitVP9Decoder): Deleted.
(WebCore::LibWebRTCProvider::registerWebKitVP8Decoder): Deleted.
* Source/WebKit/GPUProcess/GPUProcess.cpp:
(WebKit::GPUProcess::updateGPUProcessPreferences):
* Source/WebKit/GPUProcess/GPUProcess.h:
* Source/WebKit/GPUProcess/GPUProcessPreferences.cpp:
(WebKit::GPUProcessPreferences::copyEnabledWebPreferences):
* Source/WebKit/GPUProcess/GPUProcessPreferences.h:
* Source/WebKit/GPUProcess/GPUProcessPreferences.serialization.in:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::creationParameters):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::m_textAnimationController):

Canonical link: https://commits.webkit.org/287137@main
